### ISSUE TITLE
Don't fly back to current location

### ISSUE THREAD

**Issue by thany** [organicmaps/organicmaps#4047]
When I open the app, it flies to my current location. This only happens when I reopen the app after not using it for "some" amount of time.

I don't want it to fly back to my current location, I wasn't working there. I was adding bookmarks elsewhere. I cannot stop it from doing that, ever. Best I can do is be quick enugh to catch it in the act, pan the map and stop it from moving like that.

But that's a terrible workaround. I just want it NOT to use location services, when I don't need it. That's why I've been asking for the location button to be tristate between disabled, follow location, and follow location+direction.

See also #3067

**Comment by biodranik** [comment]
1. Disable location service.
2. Run OM and make sure that the location button is crossed in OM
3. Enjoy.

**Comment by claudiush** [comment]
Even though I occasionally am bothered by this behaviour the 24h timeout before resetting to current location seems like the most sensible UX compromise. I think the default should remain like that.

An alternative UX I could think of would be to combine the 24h reset to current location with a toast banner like at the bottom with a bar filling up for 5 seconds that states something like "Resetting Map to current location. Tap to cancel". So the user has a soft way to override the automatism

**Comment by biodranik** [comment]
The "counter" usually works anyway: your device needs some time to detect a location, especially if it's not connected to any network (or network location providers are disabled). Dragging the map at this moment should prevent autoscroll.
But with network location, it happens too fast... And it is suitable for most users.

**Comment by thany** [comment]
The solution is a tristate location button, just like MAPS.ME used to have since they bollocksed it up. I don't see why that's *such* a problem, tbh.

Disabling location access for the app is a hassle, because maybe I want to be able to tap the location button to make it go to my location (surely, that part makes sense, doesn't it). And maybe I want to use it again afterwards to disable it.

What is the disadvantage of a tristste button, other than users who are forgetful enough that they've tapped it to disable? That last part can happen with any setting, mind you.

**Comment by biodranik** [comment]
We used that button in MapsWithMe, but then removed the "off" switch from it, because it was inconvenient to _turn location off_ if you really need to _switch to another mode_ instead.

**Comment by thany** [comment]
How is it inconvenient? Just press the button twice. There are only three states...

If you want to switch from location&compass to location only, you press the button teice to skip over the disable state. I *really* honestly don't see the problem.

Let's put it this way: If the location button *must* also include a location&compass state, it *should* be tristate. So another solution is to split the button into two buttons:

* Follow location or don't.
* Use compass or don't.

If a tristate button really is as evil as I get the impression it might be, then it would need to be split.

**Comment by biodranik** [comment]
One additional button is even more _evil_ for the UI and UX. There should be a balance. I'm sure there are better solutions for the original issue than breaking other UX cases.

**Comment by Misalf-git** [comment]
I understand the general direction is to keep the app simple and to not pollute the settings page with options barely anyone needs.
If there would be a switch to turn off "center current location uppon app lauch", I would turn it off as well.

**Comment by thany** [comment]
If you don't want to add a button, then it merits a setting. There's no shame in adding a setting for something different people may use in different ways. It's more *evil* to try and make an app "guess" or "intelligent" as to what the user is trying to achieve.

I see three possibilities:

1. A tristate button is far easier to implement, much clearer for the user (if done well visually and such), and doesn't require an entry in the settings menu.
2. An entry in the settings menu requires the user to go into settings to change it, which will be annoying in case they find themselves oftenly have to switch it. Not the end of the world, but also no ideal, iyam.
3. To try and make this work without any option or button or whatever, is going to require that "intelligence" that some companies are madly in love with (looking at you, Google) which are ultimately confusing because it may work differently every time, there is no clear way to change it, and there is no way to predict how it'll behave next time.

I still opt for 1.

**Comment by biodranik** [comment]
Let's clarify the use case again.
1. When don't you need to "fly" to your current location? When you're adding bookmarks and want to continue doing it. Right?
2. When do you _need_ to "fly" to your current location? When you want to use the map to locate yourself.

The right questions are:
1. What is the proportion of (1) vs (2) in your case? Do you add bookmarks most of the time, and only locate yourself occasionally?
2. Are there some hints that we can detect locally on your device and use locally (without collecting data from you on some servers, in contrast with Big Tech), so the "guessing" of OM about what should be done can be improved?

**Comment by thany** [comment]
1. I prepare for my holiday (adding bookmarks and whatall) for a few weeks so then I rarely need it to fly to my current location. However, when I'm on that holiday, my markers are there and I do need it to fly to my location. Except when I'm looking around to see what the things around me were again, that I bookmarked.
2. Well, sure. Sort of. When I say "when I'm not in The Netherlands" that's gonna make for a very strange option to switch in the preferences, isn't it 😄

**Comment by biodranik** [comment]
Yes, some examples would be helpful. I know how it worked in MapsWithMe 9-10 years ago.

What I see from your description: you're not using OM to travel locally. You use it only when you're traveling somewhere.
There are two sides here:
1. OM can detect that pattern/type of user and be smart.
2. OM can become more useful locally, so you also use it more often at/near your home.

**Comment by thany** [comment]
Please don't make it smart. I said it before, and I will say it again: smartness is hard to predict and comprehend. You'd have to "blindly" trust that the app will do whatever it deems correct, without having any say in the matter.

Making it more useful locally would mean that I'm going to have to test its driving directions capability. In Google Maps it is bloody awesome, that's why I just don't switch. Also I don't "need" an offline maps app in my own country, because mobile data is only terribly expensive when abroad.

I will post some examples shortly.

**Comment by vbertola** [comment]
Yes, please, please! I hate this, I am home planning a future trip in some other part of the world, I stop for the day, I resume the following day and as soon as I open the app it starts flying back to my own home and neighbourhood, which I already know inch by inch, and I have to fight against it with my fingers, and it really doesn't want to stop!

I should just have a location button in the UI to bring me to the current location when I want so. If you do not want to change this behaviour for everyone, please give me a setting so that I can disable this annoying thing for me, forever and ever, and use the location button when I need to zoom back on my position, like Google Maps does.

Oh, and definitely it's not disabled for 24 hours - yesterday night I was working on Mexico City, I just reopened the app and it brought me back home, and only ~14 hours have passed.

**Comment by biodranik** [comment]
@rtsisyk can you please investigate the current behavior? We can "hot-fix" it by increasing the delay to 48 hours. But it looks strange to me that position mode is restored earlier.

**Comment by thany** [comment]
If it's set to 24 hours, hotfixing it to 48 hours isn't going to help. More to the point, where does this seemingly arbitrary delay of 24 hours come from? How is this decided, and why isn't this just a setting?

I've said it before and I will say it again: please don't make the app "smart". Smartness is hard to predict and oftenly comes as a surprise, making it feel unreliable. Sometimes it's good, but it's *really* hard to absolutely nail it. So we should just avoid it.

Having said that, happy new year 😀 Here's to a great year for OSM and Organic Maps.

**Comment by biodranik** [comment]
@thany Actually, there is an 8 hours limit now in the code that changes the _Not Follow_ position mode to the _Follow_ one.

For your specific case there is a working solution:
1. Disable location in the device's settings.
2. Run OM and make sure that it shows a crossed crosshair (position) button/icon.

Now if you close OM, it will save that "No Position" mode and will never restore it back until you manually press the crosshair button.

Furthermore, the search in remote areas where you are looking on the map will work better, your current position won't influence search results anymore.

**Comment by thany** [comment]
Yes, if I disable location entirely, it's """"solved"""", this is something we knew at the very start of this issue. And this workaround is completely beside the issue at hand. THe issue is to integrate this gracefully in the app, not to require 15 taps (more or less, I'm not joking) to enable/disable a seemingly simple thing.

**Comment by biodranik** [comment]
@thany I completely understand your disappointment. Please also look at it from our position: should we break the existing convenient UX for thousands of users only to satisfy your use case?

In your case, when you are using OM exclusively for planning remote trips during days or weeks, it is very easy to disable location service once, and then use OM without the current location during these days/weeks. Implementing a tri-state button only for use cases like this one looks like a big overkill to me at the moment.

**Comment by Misalf-git** [comment]
There could be more than just a few users seeing the current behavior as inconvenient but they just don't complain. Especially when there is already a lengthly discussion about it going on.
Breaking the UX for one group by drastically changing behavior is ofcourse unwanted. I believe a setting for this deserves consideration.
I haven't used too many Map Apps but quite a few don't automatically center on current location:
Vespucci, StreetComplete and Cruiser come to my mind.
OsmAnd does center on current location and I couldn't be arsed to find a setting in it's convoluted options structure to turn that off. ;)

**Comment by thany** [comment]
@biodranik No offense, but I don't think you understand. Who said we should break existing UX? A tristate button doesn't do that.

I will look at it from your perspective then: "we shouldn't change any UX because users are happy with everything the way it is". This is obviously a false statement, and laughably so. I'm half-joking of course, but it's also kind of what I'm reading in your replies. But then, bear in mind, you've actually already broken the UX by making it inconsistent.

If location can be turned ON (which you *can absolutely do*, because it can be turned off by panning away) you must also be able to turn it off with the same or similar user interaction: tapping it. So in a sense, the location button is now inconsistent:

* It can be turn OFF by panning away (which is not persisted)
* It can be turned ON by tapping it (which *is* persisted, implicitly)
* And then there's a third state, which is to follow compass.

So in a sense, the tristate-ness is already there. All I'm asking is to be able to:

* Get it to the OFF state not by panning away but (also) by tapping it. If I can tap the location button in its OFF state in order to fly back to my location, why can I not also tap it to get it back to the OFF state?
* Persist this state, even after an arbitrary number of hours, but only if OFF state was achieved by an explicit user interaction (tapping it).

The button even already has three visual states! You could add a visual cue for the other states, but you're kidding yourself if you're saying a tristate button is overkill, because it already is. It just behaves inconsistently and doesn't persist.

**Comment by biodranik** [comment]
Please don't nit-pick. Adding one more state to the current flow will definitely break UX. Example:

I want to enable map rotation to better understand my position and directions. I click my position arrow to center the map (if it is not already centered). Then I click it again to enable rotation.

After that, I want to go back to the "north up" mode. I press the location button again. Here, if you add another "GPS off" state, I will be very surprised if my position will disappear from the map.

**Comment by thany** [comment]
I'm not nitpicking.

You're making an effort to describe how the app currently works. Does it have to? It's not holy. I could copypaste your reply and describe how it would work with my solution, job done.

The challenge here lies in the UX. How can you make it so that it covers your and my use-case perfectly? tap-and-hold the location button? Make its state current much clearer than currently? Add a toaster that says "follow location is disabled, tap again to enable" or similar? All of the above?

If you keep hammering on your opinion (with all due respect) that the current location button is completely perfect, this discussion will lead nowhere.

**Comment by biodranik** [comment]
We are open to any great ideas that are 1) intuitive and 2) do not break the already working UX.

**Comment by thany** [comment]
It did it again just now. So it definitely doesn't magically remember whether it should following the current location, or just display it if it happens to be in view.

"already working UX" is a bit of a stretch, isn't it, because to me it clearly isn't working. I can't *explicitly* tell the app not to follow - but only display - my current position. It just sort of decides on its own.

I think I've given many options in the past to improve this. What do you need from me to get to a solution please?

**Comment by biodranik** [comment]
Most users expect maps to show their updated location after some time of not using the app. It helps to avoid doing a click on the position button.
Most of the other maps apps don't even try to use the logic similar to OM, they _always_ center on the current location. Breaking this behavior is a very bad idea.

It's ok to support exception cases. Coding a special setting for this case is possible, but it will require some effort to test and make sure that it won't break anything. Unfortunately, position modes logic is _very_ complicated in OM and needs some refactoring. That's why any better UX ideas are welcome.

**Comment by oleg-pah** [comment]
**Scenario**
Planning a trip to a location far from the current location, and planning over the course of multiple days.  When opening the app to continue planning, the map returns to current location, and needs to be manually returned back to the destination location.

**Problem Statement**
- When planning, the map returns to the current location, requiring to find the destination again again, adding time to planning and creating an annoying experience.
- Even when returning to the location is necessary, the flyback feature takes a long time to complete, as long as 5 seconds within the same city, which is not necessary and does not add any value to seeing the flying.

Criteria
1. In Settings, add a toggle to disables the "Follow me" functionality.  When disabled, when opening the app it will open the map to the location where it was last, and not the current location.
- Speed up the flyback feature so that it takes no longer than 1.5 to complete, no matter where in the world it is flying from.  This will reduce the annoyance factor of the feature.

**Details**
For users who use the app regularly, the most common case is to navigate from the current location to another location, so disabling the Follow Me functionality is not commonly needed.  Adding a non-commonly used feature to the main interface may not be optimal; having an option in the Settings sounds reasonable.

It may also make sense to separate out Criteria 1 and 2 into separate issues.

**Comment by biodranik** [comment]
1. The current delay until OM decides to show the current position is 30 hours. Need to check if it works properly first (e.g. if the timer resets to +30 hours after each launch before 30 hours passed, because if it's not, then it explains inconsistent behavior).
2. Faster zoom to the current position instead of a slow flyback should be fixed (slow is annoying), and it should be a separate issue.
3. As most people are planning from their home, and going somewhere from the home, adding a home/work buttons and not centering the map on the home/work area can likely solve most of the issues mentioned here even without introducing any settings.

**Comment by oleg-pah** [comment]
It would be interesting to know how many users do trip planning this way.

**Comment by thany** [comment]
Adding home/work buttons would be an example of confusing UX. I honestly, genuinely, positively cannot see why the tristate button solution is so bad. MAPS.ME had this years ago and it worked *beautifully*, until they effed it up.

> The current delay until OM decides to show the current position is 30 hours.

This feel arbitrary. Why 30 hours? Why not 24 hours? Why not a week? Why is this not a setting?

> Faster zoom to the current position instead of a slow flyback 

Should be based on the OS setting for enabled/disabled animations, but can probably be faster either way. Not particularly relevant to this issue though. Slow animation is better actually if this never gets solved, because it allows a second or so to stop the app from flying back to the current position by grabbing hold of the map "in flight".

> likely solve most of the issues mentioned here even without introducing any settings

What's so evil about adding a setting? I'm not asking for a setting, just for a user preference to be consistent and persisted. But adding a setting is not anything to be shamed about, as long as it makes sense.

**Comment by Jackalaq** [comment]
Hi all,
I've been meaning to file this as an issue for ages; I found this conversation as it captures the matter.

The suggestions, ideas and requests (most of which, I'd also had before landing here, particularly the obvious tristate button) contrast with responses that seem to assume/claim that the current set up is 'logical', 'works', 'isn't broken' and are written claiming the perspective of 'most people' (e.g. 'most people are planning from their home, and going somewhere from the home').  It doesn't read as being equally open minded on both sides and therefore seems unproductive and frustrating

For example, I use the map for non-local, off-grid planning and information referencing, whilst out and about, sometimes whilst on a train/bus or on treks/walks.  It isn't generally 'trip planning' as maps are way more than just tools to find your way from 'here to B', let alone from 'home to B'.  Since maps are such multipurpose tools, I often also use OM interchangeably for local and non local purposes.  OM really lends itself to this due to the excellent 'show as lock screen' function that I am pathetically grateful for!  This is exactly when I want to be able to toggle location (for battery saving) and centre my location without unlocking, going to settings, toggling location etc etc all whilst it's raining and I have frozen fingers!

Since offline use, locked-device use and long battery life are three of the advertised benefits of this software, I suggest that the actually current tristate button that switches to bistate mode once activated is not logical, nor properly functional.  
Current button (from location off):
🚫  -->|  ♐  <-->  🔝

Suggested:
🚫  -->  ♐  -->  🔝
  ^---------------------╯short press
Or:
🚫  -->|  ♐  <-->  🔝
  ^----------╯<-------╯long press to identify rarer case use

As @biodranik says, if the user's location marker disappears from the map whilst cycling through the 3 states, then they will already have their finger on the right button to correct this and will know that this is the 'locate me button', then I don't see this logical tristate functionality breaking the UX that currently feels broken.  Instead, it is a fix!  Nevertheless, I would be happy with long press to deactivate location and,  way, would use it extensively.

Finally, I would like to highlight that I do turn off location - **just to save battery** - during the simple case use where I am only using the map to navigate.  I don't want GPS warming my pocket as it tries to get signal whilst I know where I am.  In this case the UX is definitely broken as the location button only turns on, not off as predicted.  This must be a very common use case and is surely indisputable.  Rapid toggling saves battery and suits the wet, cold finger situation given above!

Lets remember that most users don't make it to GitHub.  Thanks for reading and considering.

(For interest, attached is a tristate software button screenshot from /e/OS volume control. The tributton appears on pressing the monobutton)
![Screenshot_20240828-155925_Firefox](https://github.com/user-attachments/assets/f1b62868-062d-403e-a5dc-ef33f4d35c53)

**Comment by biodranik** [comment]
Are you aware of other map apps that support turning GPS off using the "locate me"/"enable map rotation" button? 

We implemented it in MapsWithMe, but then that behavior was changed, mainly because the most cases were switching from "rotate" to "north up", and doing it in one click instead of turning GPS off and then switching to "north up" is a better UX.

**Comment by Aurareus** [comment]
I think that argument is totally sound and reasonable; most users probably don't need the functionality to be able to toggle the "locate me" button off, so being able to switch between its two enabled modes with one tap instead of two is preferable, as stated.

Some users do require that functionality though (the ability to toggle the "locate me" button off), as is evidenced by the fact that multiple people have expressed that need in this issue.

So, what is the specific reasoning for why the proposed solution of long-pressing (tap and hold) to toggle the "locate me" button to off is bad UX?
It seemingly simultaneously serves the users who require the functionality to toggle it off whilst not removing the average users ability to toggle between the two enabled modes with a single tap.

**Comment by oleg-rswll** [comment]
Currently the button is used for map perspective or position.  If adding a 3rd view to toggle through, this may be a reasonable place to have it.  The question is: would adding this view create complexity or confusion for the vast majority of people who do not need to turn off flyback?

A long press to turn off Flyback, seems reasonable, it doesn't affect any existing functionality and adds functionality for those who are aware of it.  The only potential issue may be, since this is a "hidden" feature, people may accidentally long press on the button, turn off the fly back, and not know how to turn it back on.

**Comment by Atemu** [comment]
Perhaps I'm missing something but why does the app need to decide whether to fly the user back to their position at all?

The location button also already is a tri-state to encode this exact preference and I find using it to be quite intuitive:

You press on it once to jump to your location, you press again to make it rotate with your compass.  
To make it stop following your location and look at something else instead, you simply drag the map to the desired location. The button automatically switches to the greyed-out "don't follow"/"look around" icon.

When the app re-opens, it should simply persist that same state. Unconditionally. No magic/smartness required.

If you were in the "look around" mode before closing the app, you should simply be restored to what you were previously looking at when re-opening the app, no matter how much time has passed.

If you then actually wanted to look at your current location, you'd simply press the location button (that is currently grey to indicate that you're in "look around" mode) and thus initiate a "fly to current location".  
No need for the app to decide whether you want to do that for you because it only requires one press of an obvious button to make it happen yourself, should you desire it.

**Comment by thany** [comment]
I would like to add, @Atemu, that if a button has a tristate at all (which it does), logic dictates that it must be possible to *explicitly* trigger that third state. Tapping the button should switch between the three states, simply because it has three states. That's all.

To illustrate,

1. A user might have panned the map without realising that button is even there at all
2. Then, they tap the button to fly back to their current location, just to explore what that button does.
3. They then realise it's not what they wanted, and want to put the button back to its previous (third) state.
4. They can't, no matter how many times they tap. Result: confusion.

And yes, I know. It's convention and all that. Google Maps has it too. So what? Google Maps has a lot of crappy features some folks don't want, and prefer another app because of it.

At the same time, I can be convinced. @biodranik, show me some solid research or survey results that users would like a three-state location button to switch between only two states on tapping. I'm sure this behaviour wasn't changed on a whim? Because need I remind you, it used to work (as I see it) correctly.

**Comment by Atemu** [comment]
I fully get what you're saying but you also have to consider that toggling between follow compass and only follow is going to be the most common thing happening while navigating. Having it also stop following you could be quite confusing and you might not immediately notice that happened because you might be standing still at that moment (quite likely actually). Dragging the map OTOH is quite an unambiguous signal: "I want to look exactly here and stop following me".    
I think I'd prefer an actual tri-state too but I could see the argument being made the other way too to the point where I think you'd need to do an actual randomised trial to figure out which is better from a UX perspective.

That's only relevant tangentially to this issue though and I think there are much more pressing issues that need to be addressed first/resources be spent on them.  
Google isn't the best standard for UX but I'd expect something like this to have already gone through internal UX testing and if they can get away with that UX, so can OM. As mentioned, I'd prefer it to be different too but you have to concede that this isn't a very pressing issue.

What's actually important for this issue (which I do think is quite a bit more significant) is that the state encoded by the tri-state must never be changed by the app, no matter how "smart" the decision when and how to change it may be because it is unpredictable and even if it was predictable, often simply not desirable.

**Comment by pastk** [comment]
This issue became very long with a lot of discussions.

If there was a consensus reached then it'd be great to write a concise short summary (what needs to be solved and how to solve it) (and create new (sub-)issues if necessary). I.e. make it more actionable 🙏
